### DISCUSSION TITLE
Questions relative to porting of the HA OS to Beagle Bone Black

### DISCUSSION THREAD

**Discussion by Didifred** [home-assistant/operating-system#3644] [Q&A] [answered]
I'm quite advanced in porting HA OS to Beagle Bone Black, i have several questions before to do a pull requests if the community is interested ...
- It seems mandatory ( initialization fails in others cases) to have packages BR2_PACKAGE_RPI_RF_MOD and BR2_PACKAGE_RPI_RF_MOD_DTS, it shall concerns obviously only RPI hw ? 
- I have one warning at initialization of Home Assistant application that concerns pulsaudio, in *_defconfig the package is not compiled (BR2_PACKAGE_PULSEAUDIO). It sounds every install has this warning so ?
- I've setup BR2_PACKAGE_HASSIO_MACHINE="tinker" and BR2_PACKAGE_OS_AGENT_BOARD="Tinker", since i need something close. Why the supervisor and os agent are dependant to the board BSP, naively i was thinking that HA OS provided APIs independant of the board ? Using it for beagle bone seems work ok, it do not appear to be dependant of board in fact.
- In order to flash the EMMC, i simply dd the image from a SD card  bootup debian 12 ( the haos-data expansion is about 3G and the OS itself about 450M shrinking the partition mapping a bit, i'm near to the 3.6G available ...). Then i move data to external media (SD or SSD) via HA web application. Running OS on EMMC give reasonable starting time of the application comparing SD. Maybe i can even remove the haos-data partition from the image, flash it in EMMC and burn the data.ext4 on external media, and first start like this ? Does OS will scan automatically for the partition in external media at first start ?

**Comment by sairon** [comment]
Hey, sorry for replying this late but maybe my answers would still have some use :)

> It seems mandatory ( initialization fails in others cases) to have packages BR2_PACKAGE_RPI_RF_MOD and BR2_PACKAGE_RPI_RF_MOD_DTS, it shall concerns obviously only RPI hw ?

It shouldn't be needed - while it's not exclusively for RPi HW (there's support for some more boards), some of our targets don't use it at all, e.g. Green. I'm not sure what you mean by failed initialization.

> I have one warning at initialization of Home Assistant application that concerns pulsaudio, in *_defconfig the package is not compiled (BR2_PACKAGE_PULSEAUDIO). It sounds every install has this warning so ?

Again, not sure what warning you mean but PulseAudio is normally running in a separate container (https://github.com/home-assistant/plugin-audio), it shouldn't be needed to have it in the base OS.

> I've setup BR2_PACKAGE_HASSIO_MACHINE="tinker" and BR2_PACKAGE_OS_AGENT_BOARD="Tinker", since i need something close. Why the supervisor and os agent are dependant to the board BSP, naively i was thinking that HA OS provided APIs independant of the board ? Using it for beagle bone seems work ok, it do not appear to be dependant of board in fact.

OS Agent has some device-specific features that are depending on a [compile-time flag](https://github.com/home-assistant/os-agent/blob/main/main.go#L29). Maybe it is something that could be done differently but right now it's a historical decision that needs to be taken as a fact. Similarly, `HASSIO_MACHINE` is used to determine what Docker images would be used by the Supervisor, which may have some specific extra packages installed, etc.

> Maybe i can even remove the haos-data partition from the image, flash it in EMMC and burn the data.ext4 on external media, and first start like this ? Does OS will scan automatically for the partition in external media at first start ?

No need to remove anything from the image. Simply use the data disk feature and the eMMC will be used only for the other partitions: https://www.home-assistant.io/common-tasks/os/#using-external-data-disk

**Reply by Didifred** [reply, marked as answer]
Hi, i've seen the answers some weeks ago, here my answers ::
1 - Green perhaps doesn't not use it, but in actual green_defconfig, we find BR2_PACKAGE_RPI_RF_MOD=y.
     I mean by fails at initialization, the initialization of the services at os initialization, no more sure exactly at which stage since the time ...

2 - I mean here on first start on the web browser, the splash screen saying "please waiting" and we can open a log window. There is a yellow line indicating pulseaudio problem :
WARNING (SyncWorker_1) [supervisor.host.sound] Can't update PulseAudio data: Failed to connect to pulseaudio server
EDIT : at least also on RPI3

3- I see effectively leds of the board yellow and green, as an architecture point of view it is strange to implement it in the OS agent. The role of it is from the repo is " ... allows the Home Assistant Supervisor to communicate with the host operating system".
I can understand the particularities of specific platform, but nevertheless when porting to other platform need to create new docker. On my porting i 've used BR2_PACKAGE_HASSIO_MACHINE="tinker" and BR2_PACKAGE_OS_AGENT_BOARD="Tinker" and HA is running like a charm, that's make me thinking the architecture in near to be generic yet.
May be to create a new BR2_PACKAGE_OS_AGENT_BOARD="GenericArmv7" ?

4- I need to remove the data partittion because the image do not fit on EMMC after first dockers installs.
If i remember well i've manually copied the data partition image in SD card (i use SD card as a external storage), and it has been found from OS booting from EMMC.

Last question : are you interested to add beagle board black support ? In this case i can do a pull request ?
My fork is there : https://github.com/Didifred/operating-system-beaglebone

**Comment by avi951** [comment]
@Didifred, could you guide me on the partition table?
I have tried to port my custom board to HAOS, I am able to make only two partition as I am using `genimage.cfg`.

```
image sdcard.img {
	hdimage {
		partition-table-type = "gpt"
	}

	partition uboot {
		in-partition-table = "false"
		image = "idbloader.img"
		offset = 32K
	}

	partition uboot {
		in-partition-table = "false"
		image = "u-boot.itb"
		offset = 8M
	}

	partition uboot-env {
		partition-type-uuid = 3de21764-95bd-54bd-a5c3-4abe786f38a8
		offset = 16M
		size = 16M
	}

	partition rootfs {
		partition-type-uuid = L
		image = "rootfs.ext4"
		bootable = true
	}
}
```

**Reply by Didifred** [reply]
What is your target ? Which partitions are missing ? Gen image is for GPT disk.

**Reply by avi951** [reply]
@Didifred, I am using Asus ``tinker_defconfig`` for deployed docker things...
And For Partition, I think HAOS has fixed partition with specific value, like for ``kernel`` **24 MB** and for ``system`` **256 MB**. 
Image is built properly, but somehow ``idbloader`` and ``u-boot.itb are not properly aligned with the GPT partition table and throws below error :

```
DDR 73dffea49e typ 23/09/20-14:18:32,fwver: v1.14
LPDDR4X, 2112MHz
channel[0] BW=16 Col=10 Bk=8 CS0 Row=16 CS1 Row=16 CS=2 Die BW=16 Size=2048MB
channel[1] BW=16 Col=10 Bk=8 CS0 Row=16 CS1 Row=16 CS=2 Die BW=16 Size=2048MB
channel[2] BW=16 Col=10 Bk=8 CS0 Row=16 CS1 Row=16 CS=2 Die BW=16 Size=2048MB
channel[3] BW=16 Col=10 Bk=8 CS0 Row=16 CS1 Row=16 CS=2 Die BW=16 Size=2048MB
Manufacturer ID:0x1
CH0 RX Vref:28.4%, TX Vref:19.8%,19.8%
CH1 RX Vref:25.4%, TX Vref:20.8%,19.8%
CH2 RX Vref:29.3%, TX Vref:18.8%,18.8%
CH3 RX Vref:26.7%, TX Vref:19.8%,19.8%
change to F1: 528MHz
change to F2: 1068MHz
change to F3: 1560MHz
change to F0: 2112MHz
out
U-Boot SPL board init
U-Boot SPL 2017.09-ge4e124926e-230922 #lxh (Sep 25 2023 - 10:58:38), fwver: v1.13
unknown raw ID 0 0 0
unrecognized JEDEC id bytes: 00, 00, 00
Trying to boot from MMC2
No misc partition
sys malloc pool space exhausted
alloc_read_gpt_entries: ERROR: Can't allocate lX bytes for GPT Entries
GPT: Failed to allocate memory for PTE
part_get_info_efi: *** ERROR: Invalid GPT ***
part_get_info_efi: *** ERROR: Invalid Backup GPT ***
spl: partition error
```

**image-spl-spl.cfg**
```
image spl.img {
	size = ${BOOT_SPL_SIZE}

	hdimage {
		partition-table-type = "gpt"
		fill = "yes"
	}

	partition idbloader{
		in-partition-table = "no"
		image = "idbloader.img"
		offset = 32K
	}

	partition u-boot-itb {
		in-partition-table = "no"
		image = "u-boot.itb"
		offset = 8M
		size = 4M
	}
}
```
**meta** file
```
BOARD_ID=tinker
BOARD_NAME="Board"
CHASSIS=embedded
BOOTLOADER=uboot
KERNEL_FILE=Image
PARTITION_TABLE_TYPE=gpt
BOOT_SIZE=24M
BOOT_SPL=true
BOOT_SPL_SIZE=16M
BOOT_ENV_SIZE=0x8000
SUPERVISOR_MACHINE=tinker
SUPERVISOR_ARCH=armv7
```

After flashing the Image, here is the problem...
Build Image size : 2.1 GB
SD Card Size : 16 GB

```
sudo fdisk  -l /dev/sdc
GPT PMBR size mismatch (4194303 != 31116287) will be corrected by write.
The backup GPT table is not at the end of the device.
Disk /dev/sdc: 14.84 GiB, 15931539456 bytes, 31116288 sectors
Disk model: Storage Device
Units: sectors of 1 * 512 = 512 bytes
Sector size (logical/physical): 512 bytes / 512 bytes
I/O size (minimum/optimal): 512 bytes / 512 bytes
Disklabel type: gpt
Disk identifier: 5CE7B822-3F3E-4DFA-B31D-3FCD78F02D04

Device       Start     End Sectors  Size Type
/dev/sdc1    34816   83967   49152   24M EFI System
/dev/sdc2    83968  133119   49152   24M Linux filesystem
/dev/sdc3   133120  657407  524288  256M Linux filesystem
/dev/sdc4   657408  706559   49152   24M Linux filesystem
/dev/sdc5   706560 1230847  524288  256M Linux filesystem
/dev/sdc6  1230848 1247231   16384    8M Linux filesystem
/dev/sdc7  1247232 1443839  196608   96M Linux filesystem
/dev/sdc8  1443840 4065279 2621440  1.3G Linux filesystem
```
